### PR TITLE
Feature/ab#1044 fix for kics

### PR DESCRIPTION
**Pull request by Anthony-Mckale**
Fixes for kics issues

background: kics binaries are no longer available, solution: extracted kics from kics debian docker image (and locked to specific version to prevent unexpected breakages)

Work done:
DOCKER BUGS
- extracted kics from kics debian docker image (and locked to specific version to prevent unexpected breakages)

DOCKER IMPROVEMENTS
- updated base image to include dot net libraries
   (now based off "mcr.microsoft.com/dotnet/sdk:6.0-bullseye-slim")
- combined some of our docker layers according to best pracice

IMPROVEMENTS
- reorganised our utils into sub folders
- added PluginMeta to abstract away code based version checking / and simplification of Tool/ReporterPlugins

IMPROVEMENTS SPEED UPS
- renamed WINDOWS_DOCKER_WORKAROUND to USE_SOURCE_COPY, after finding 80% plus gains even on linux systems.
  also turned on as default due to speed increases
- added USE_SOURCE_COPY to trufflehog
- added USE_SOURCE_COPY to kics (for mono repos, massive speed increases 800s+ for eze-console to 44s, by skipping node_modules [unfortunately using kics -e exclude flag wasn't efficient in ignoring these folders])

BUGS
- bug fix for eze-cli\eze\utils\io\file_scanner.py cache_workspace_into_tmp()
  wasn't working with non cmd windows volumes

DOCS
- added caching sharing to read me
  (for speeding up SCA tooling typically)
- updated eze-dev image, so can be used by developers to test linux performance

BAD
- some dotnet artefacts from 1045 dotnet support ticket, it's incoming soon within sprint

### CODE REVIEW

File: eze/utils/cli/run.py [hunk 1]
```diff
@@ -143,7 +142,7 @@ def _append_to_list(command_list: list, appendees, config: dict) -> str:
     return command_list
 
 
-def _append_multi_value_to_list(command_list: list, appendees, config: dict) -> str:
+def _append_multi_value_to_list(command_list: list, appendees, config: dict) -> list:
     """annotate command string with appendees which are "dict args=config-key kv" or "list of config-keys" """
```
Comment (mgonzales-rs) [inline eze/utils/cli/run.py:146]: same comments for functions: _append_to_list   ...multi_value_to_list    _append_short_flags_to_list

File: tests/utils/cli/test_run.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+def test_extract_executable__safety_std():
+    expected_output = "safety"
+    test_input = "safety check --full-report --api=1234 -r something/requirements.txt -r something-else/requirements.txt --json --output /tmp/something-temp something-at end"
+    output = _extract_executable(test_input)
+    assert output == expected_output
+
+
+def test_extract_executable__trufflehog_std():
+    expected_output = "trufflehog3"
+    test_input = "trufflehog3  -f json amplify public src scripts .ezerc.toml package.json -o /tmp/.eze-temp/tmp-truffleHog-report.json --exclude node_modules/.* #current-cloud-backend/.* backend/function/ezemcscanresult/src/node_modules/.* backend/awscloudformation/.*': 'trufflehog3  -f json amplify public src scripts .ezerc package.json -o /tmp/.eze-temp/tmp-truffleHog-report.json --exclude node_modules/.* #current-cloud-backend/.* backend/function/ezemcscanresult/src/node_modules/.* backend/awscloudformation/.*"
```
Comment (mgonzales-rs) [inline tests/utils/cli/test_run.py:50]: is there a particular reason why are we using the same command 2 times?

File: tests/plugins/tools/test_semgrep.py [hunk 3]
```diff
@@ -89,52 +87,59 @@ def test_creation__with_config(self):
...
-        input_config = {"ADDITIONAL_ARGUMENTS": "--something foo", "REPORT_FILE": "foo_report.json"}
+        input_config = {
+            "ADDITIONAL_ARGUMENTS": "--something foo",
+            "REPORT_FILE": "foo_report.json",
+            "USE_SOURCE_COPY": False,
+        }
+        absolute_report = create_absolute_path(input_config["REPORT_FILE"])
 
-        expected_cmd = "semgrep --optimizations all --json --time --disable-metrics -q --use-git-ignore -c p/ci -c p/dockerfile -c p/nodejs -c p/javascript -o foo_report.json --something foo"
+        expected_cmd = f"semgrep --optimizations all --json --time --disable-metrics -q --use-git-ignore -c p/ci -c p/dockerfile -c p/nodejs -c p/javascript -o '{absolute_report}' --exclude 'test_*.py' --exclude '*.test.js' --exclude tests --exclude __tests__ --something foo"
```
Comment (mgonzales-rs) [inline tests/plugins/tools/test_semgrep.py:135]: does semgrep needs the paths to exclude   as   --exclude {path} --exclude {path}?

**Review by mgonzales-rs — CHANGES_REQUESTED** [review]
left some minor comments, testing it locally is giving me an error.

File: Dockerfile [hunk 4]
```diff
@@ -34,140 +35,115 @@
...
+    DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 \
+    # dotnet env, add tools into PATH
+    PATH="$PATH:/home/ezeuser/.dotnet/tools" \
+    # kics env, add kics into PATH
+    PATH="$PATH:/app/bin"
+
+# apt-get installs
+# nosemgrep
 RUN apt-get update \
+    && mkdir -p /usr/share/man/man1 /usr/share/man/man2 \
```
[the remark this answers is not in the document]
Comment (Anthony-Mckale) [reply Dockerfile:81]: +1 updated in 1015

File: Dockerfile [hunk 5]
```diff
@@ -34,140 +35,115 @@
...
-    && gitleaks --version \
-    && echo "SIZETAG:Tool:gitleaks"
+    && gitleaks --version
 
-#
-## Install Kics tools
-RUN set -o pipefail \
-    && curl -sSfL https://raw.githubusercontent.com/Checkmarx/kics/master/install.sh | sh -s -- -b /usr/local/bin \
-    && echo "SIZETAG:Tool:checkmarx/kics"
+# Add kics binaries from offical image
```
[the remark this answers is not in the document]
Comment (Anthony-Mckale) [reply Dockerfile:133]: updated in 1015

File: eze/utils/cli/run.py (hunk 1 above)
Comment (Anthony-Mckale) [reply eze/utils/cli/run.py:146]: updated in 1045

File: tests/utils/cli/test_run.py (hunk 2 above)
Comment (Anthony-Mckale) [reply tests/utils/cli/test_run.py:50]: it's previously written test case

basically it's all different types of extraction

but test_extract_executable__safety_std / test_extract_executable__trufflehog_std but will remove teh surplus "test_extract_executable__trufflehog_std" isn't kinda the same as "test_extract_executable__safety_std"

in 1045

File: tests/plugins/tools/test_semgrep.py (hunk 3 above)
Comment (Anthony-Mckale) [reply tests/plugins/tools/test_semgrep.py:135]: that's just the way semgrep works

some tools like "--arg value1 value2"
some tools like "--arg value1 --arg value2"

**Review by mgonzales-rs — APPROVED** [review]

File: tests/utils/cli/test_run.py (hunk 2 above)
Comment (mgonzales-rs) [reply tests/utils/cli/test_run.py:50]: yes I saw it was previously in the the old test case so I wanted to know the purpose, thanks :)

**Pull request merged**